Links, paths, and anchors (#453)
holochain/docs-pages: src/pages/build/links-paths-and-anchors.md

```rust
use hdk::hash_path::path::*;
use movie_integrity::*;

// This will create a two-level path that looks like:
//    "movies_by_first_letter" → "g"
let path_to_movies_starting_with_g = Path::from("movies_by_first_letter.g")
  // A path requires a link type that you've defined in the integrity zome.
  // Here, we're using the `MovieByFirstLetterAnchor` type that we created.
  .typed(LinkTypes::MovieByFirstLetterAnchor)?;

// Make sure it exists before attaching links to it -- if it already exists,
// ensure() will have no effect.
path_to_movies_starting_with_g.ensure()?;

let create_link_hash = create_link(
    path_to_movies_starting_with_g.path_entry_hash()?,
    movie_entry_hash,
    LinkTypes::MovieByFirstLetter,
    ()
)?;
```

Retrieve all the links on a path by constructing the path, then calling `get_links` with its hash and link type:

```rust
use hdk::hash_path::path::*;
use movie_integrity::*;

let path_to_movies_starting_with_g = Path::from("movies_by_first_letter.g");

> let links_to_movies_starting_with_g = get_links(

Reviewer: ```suggestion

```

let links_to_movies_starting_with_g = get_links(
    // A path doesn't need to have a type in order to compute its hash.
    GetLinksBuilder::try_new(path_to_movies_starting_with_g.path_entry_hash()?, LinkTypes::MovieByFirstLetter)?
)?;

> )?;

Reviewer: ```suggestion
```

A piece of content itself can contain a hash in one of its fields, and that's great for modelling a _many-to-one relationship_. For instance, an [**action**](/build/#entries-actions-and-records-primary-data), one of the primary types of data in the database, points to the action that precedes it and to the entry it creates, updates, or delete.

And if the number of things the content points to is small and doesn't change often, you can model a _many-to-many relationship_ using a field that contains an array of hashes. But at a certain point this becomes hard to manage, especially if that list grows or regularly changes.

Holochain's hash table stores _metadata_ in addition to primary addressable content. This lets you attach **links** to an address in the database. You can then retrieve a full or filtered list of links from that address in order to discover more primary content. In this way you can build up a fully traversable graph database.

### Define a link type

> Every link has a type that you define in an integrity zome, just like [an entry](/build/entries/#define-an-entry-type). Links are simple enough that they have no entry content. Instead, their data is completely contained in the actions that write them. Here's what a link creation action contains, in addition to the [common action fields](/build/working-with-data/#entries-actions-and-records-primary-data):

Reviewer: The phrasing of sentences 2-4 is a bit confusing to me. I might just say explicitly: "a link is commited to the source chain as an action with no associated entry"

* A **base**, which is the address the link is attached to and _points from_
* A **target**, which is the address the link _points to_
* A **type**

> * An optional **tag** that can hold a small amount of arbitrary bytes, up to 4 kb

Reviewer: hdk docs say the max is 1 kb (https://docs.rs/hdk/latest/hdk/link/fn.create_link.html) -- I can't actually find where this limit is applied in the codebase so not sure which is accurate

Author: yup, you're right: https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/sys_validate.rs#L34

  MovieByFirstLetter,
  // other types...
}
```

## Create a link

As with entries, you'll normally want to store your link CRUD code in a [**coordinator zome**](/resources/glossary/#coordinator-zome). You can read about why in the page on [entries](/build/entries/#create-an-entry).

Create a link by calling [`hdk::prelude::create_link`](https://docs.rs/hdk/latest/hdk/link/fn.create_link.html). If you used the `hdk_link_types` macro in your integrity zome (see [Define a link type](#define-a-link-type)), you can use the link types enum you defined, and the link will have the correct integrity zome and link type indexes added to it.

```rust
use hdk::prelude::*;
// Import the link types enum defined in the integrity zome.
use movie_integrity::*;

let director_entry_hash = EntryHash::from_raw_36(vec![ /* bytes of the hash of the Sergio Leone entry */ ]);
let movie_entry_hash = EntryHash::from_raw_36(vec![ /* bytes of the hash of the Good, Bad, and Ugly entry */ ]);

let create_link_action_hash = create_link(
  director_entry_hash,
  movie_entry_hash,
  LinkTypes::DirectorToMovie,
  // Create an optional search index value for fast lookup.
  vec!["year:1966".as_bytes()].into()
);
```

> Links can't be updated; they can only be created or deleted.

Reviewer: Should we give some clarifying rationale for why they can't be deleted?

Author: sure... if only I knew why 🤣

## Delete a link

Delete a link by calling [`hdk::prelude::delete_link`](https://docs.rs/hdk/latest/hdk/link/fn.delete_link.html) with the create-link action's hash.

```rust
use hdk::prelude::*;

let delete_link_action_hash = delete_link(
  create_link_action_hash
);
```

> A link is considered dead once its creation action has one or more delete-link actions associated with it.

Reviewer: I might also copy the same disclaimer from the entries page: "deleting" a link just creates another action marking the link as deleted and does not actually delete any data. And that a future potential purge feature will not apply to links as they are only actions.

> A link is considered dead once its creation action has one or more delete-link actions associated with it.

Reviewer: I'm not sure about using the term "dead" here -- in the web 2 context it usually means that a link's target is no longer hosted anywhere. I haven't seen the term used elsewhere in holochain, maybe we should just stick with "deleted"?

Author: I think I'll stick with 'dead' for now -- I get that it has a different meaning in Web2, but it's used in the [codebase](https://github.com/holochain/holochain/blob/develop/crates/holochain_zome_types/src/metadata.rs#L79-L80) to describe entries whose creation actions have all been marked deleted, and I want some word to suggest that it hasn't been removed -- there's still a body lying around :)

Reviewer: love the clarification around dead/deleted you added.

> ## Getting hashes for use in linking

Reviewer: I wonder if it would be clearer / better mental separation to move this section onto its own page "Hashes", and then keep this page only about links / anchors / paths.

Author: yup, great thing to do in a follow-up -- or put all the content on 'Working with Data'

Author: done!

Because linking is all about connecting hashes to other hashes, here's how you get a hash for a piece of content.

> !!! info A note on the existence of data

Reviewer: The section title could be more descriptive. Maybe "Link Addresses May Not Have Content"?

!!!

### Actions

Any host function that records an action on an agent's source chain, such as `create`, `update`, `delete`, `create_link`, and `delete_link`, returns the hash of the action. You can use this for further writes in the same function call, or you can return it to the client that called the function so that it can use it later. You can also [pass it to another function](/concepts/8_calls_capabilities/) --- either one in the same cell, another cell in the agent's hApp, or another cell in the same network.

<!-- TODO: remove/simplify this with a pointer to the lifecycle document when I write it -->
!!! info Action hashes aren't certain until zome function lifecycle completes
When you get an action hash back from the host function that creates it, it doesn't mean the action is available on the DHT yet, or will ever be available. The action isn't written until the function that writes it completes, then passes the action to validation. If the function or the validation fail, the action will be discarded. And if it is successful, the action won't become fully available on the DHT until it's been published to a sufficient number of peers.

> It's safer to share action hashes with other peers or cells in a callback called `post_commit()`. If your coordinator zome defines this callback, it'll be called after every successful function call within that zome.

Reviewer: Yeah this confused me too. In the context of linking to/from the action within the same zome call it was created there's no issue since all the actions will either be committed to the source chain or discarded atomically

```rust
if let Action::Update(action_data) = action {
  let replaced_entry_hash = action_data.original_entry_address;
} else if let Action::Delete(action_data) = action {
  let deleted_entry_hash = action_data.deletes_entry_address;
}
```

### Agent

An agent's ID is just their public key, and an entry for their ID is stored on the DHT. The hashing function for an agent ID entry is just the literal value of the entry. This is a roundabout way of saying that you link to or from an agent using their public key as a hash.

An agent can get their own ID by calling [`hdk::prelude::agent_info`](https://docs.rs/hdk/latest/hdk/info/fn.agent_info.html). Note that agents may change their ID if their public key has been lost or stolen, so they may have more than one ID over the course of their participation in a network.

```rust
use hdk::prelude::*;

let my_first_id = agent_info()?.agent_initial_pubkey;
let my_current_id = agent_info()?.agent_latest_pubkey;
```

All actions have their author's ID as a field. You can get this field by calling the action's `author()` method:

```rust
let author_id = action.author();
```

> ### External reference

Reviewer: May be worth noting that the external hash still needs to fit in 32 bytes.

Because an external reference comes from outside of a DHT, it's up to you to decide how to get it into the application. Typically, an external client such as a UI or bridging service would pass this value into your app.

```rust
use hdk::prelude::*;
use movie_integrity::*;

#[hdk_extern]
fn add_movie_poster_from_ipfs(movie_entry_hash: EntryHash, ipfs_hash_bytes: Vec<u8>) {
  let ipfs_hash = ExternalHash::from_raw_32(ipfs_hash_bytes);
  create_link(
    movie_entry_hash,
    ipfs_hash,
    LinkTypes::IpfsMoviePoster,
    ()
  );
}
```

### DNA

There is one global hash that everyone knows, and that's the hash of the DNA itself. You can get it by calling [`hdk::prelude::dna_info`](https://docs.rs/hdk/latest/hdk/info/fn.dna_info.html).

```rust
use hdk::prelude::*;

let dna_hash = dna_info()?.hash;
```

> !!! info Linking from a DNA hash is not recommended

Reviewer: I think the conceptual explanation of "hot spots" deserves its own section on this page, rather than hiding it in the DNA hash section. And in general I think it's okay to create links from a known base as long as there is some reasonable expected limitation to the number of links that will be created off it.

!!!

## Retrieve links

Get all the live links attached to a hash with the [`hdk::prelude::get_links`](https://docs.rs/hdk/latest/hdk/link/fn.get_links.html) function.

```rust
use hdk::prelude::*;
use movie_integrity::*;

let director_entry_hash = EntryHash::from_raw_36(vec![/* Sergio Leone's hash */]);
let movies_by_director = get_links(
  director_entry_hash,
  LinkTypes::DirectorToMovie,
  None
)?;
let movie_entry_hashes = movies_by_director
  .iter()
  .map(|link| link.target.into_entry_hash()?);
```

If you want to filter the returned links by tag, pass some bytes as the third parameter. The peer holding the links for the requested base address will return a list of links whose tag starts with those bytes.

```rust
use hdk::prelude::*;
use movie_integrity::*;

let movies_in_1960s_by_director = get_links(
  director_entry_hash,
  LinkTypes::DirectorToMovie,
  Some(vec!["year:196".as_bytes()].into())
);
```

> To get all live _and dead_ links, along with any deletion actions, use [`hdk::prelude::get_link_details`](https://docs.rs/hdk/latest/hdk/link/fn.get_link_details.html). This function has the same options as `get_links`

Reviewer: same note as above about the terms "dead" and "live"

```rust
use hdk::prelude::*;
use movie_integrity::*;

let movies_plus_deleted = get_link_details(
  director_entry_hash,
  LinkTypes::DirectorToMovie,
  None
);
```

### Count links

> If all you need is a count of matching links, such as for an unread messages badge, use [`hdk::prelude::count_links`](https://docs.rs/hdk/latest/hdk/prelude/fn.count_links.html). It has a different input with more options for querying (we'll likely update the inputs of `get_links` and `count_links` to match `count_links` in the future).

Reviewer: Might be worth noting some gotchas with count_links: it goes to the network first, so if you end up asking 2 different peers with differing dht syncs you may get 2 different answers (which may be odd behavior if your unread messages badge shows a smaller number than before without you reading any messages). You may also get a different count as the number returned from get_links for the same reason.

Author: Good point, although I feel like there's an entire story about that re: eventual consistency everywhere.

Reviewer: Yeah it's more than just that though since it doesn't go to your cache so you may have more links than you receive via count

```rust
use hdk::prelude::*;
use movie_integrity::*;

let my_current_id = agent_info()?.agent_latest_pubkey;
let today = sys_time()?;
let number_of_reviews_written_by_me_in_last_month = count_links(
  LinkQuery::new(movie_entry_hash, LinkTypes::MovieReview)
    .after(Some(today - 1000 * 1000 * 60 * 60 * 24 * 30))
    .before(Some(today))
    .author(my_current_id)
);
```

## Paths and anchors

> Sometimes the easiest way to discover a link base is to embed it into the application's code. You can create an **anchor**, an entry whose content is a well-known blob, and hash that blob any time you need to retrieve links. This can be used to simulate collections or tables in your graph database. As [mentioned](#getting-hashes-for-use-in-linking), the entry does not even need to be stored; you can simply create it, hash it, and use the hash in your link.

Reviewer: I think calling it an entry or blob that doesn't need to exist just adds confusion. An anchor as a hash that is already known to every user. One way to make it known to every user is to hard-code it. It doesn't need to be generated by hashing some human meaningful data, it could just be 32 arbitrary bytes.

Reviewer: It's also a little confusing that we have this *concept* of an anchor, and then we also have an opinionated implementation in the HDK - might be helpful to distinguish them explicitly

Author: Thoroughly agree re: concept vs named implementation.

While you can build this yourself, this is such a common pattern that the HDK implements it for you in the [`hdk::hash_path`](https://docs.rs/hdk/latest/hdk/hash_path/index.html) module. The implementation supports both anchors and **paths**, which are hierarchies of anchors.

> !!! info Avoiding DHT hot spots

Reviewer: same note as above -- I think hot spots could be expanded on in its own section

  path_to_movies_starting_with_g.path_entry_hash()?,
  movie_entry_hash,
  LinkTypes::MovieByFirstLetter,
  ()
)?;
```

Retrieve all the links on a path by constructing the path, then getting its hash:

```rust
use hdk::hash_path::path::*;
use movie_integrity::*;

// Note that a path doesn't need to be typed in order to compute its hash.
let path_to_movies_starting_with_g = Path::from("movies_by_first_letter.g");
let links_to_movies_starting_with_g = get_links(
  path_to_movies_starting_with_g.path_entry_hash()?,
  LinkTypes::MovieByFirstLetter,
  None
)?;
```

Retrieve all child paths of a path by constructing the parent path, typing it, and calling its `children_paths()` method:

```rust
use hdk::hash_path::path::*;
use movie_integrity::*;

let parent_path = Path::from("movies_by_first_letter")
  .typed(LinkTypes::MovieByFirstLetterAnchor);
let all_first_letter_paths = parent_path.children_paths()?;
// Do something with the paths. Note: this would be expensive to do in practice.
let links_to_all_movies = all_first_letter_paths
  .iter()
  .map(|path| get_links(path.path_entry_hash()?, LinkTypes::MovieByFirstLetter, None)?)
  .flatten()
  .collect();
```

### Anchors

> In the HDK, an 'anchor' is just a path with two levels of hierarchy. The examples below show how to implement the path-based examples above, but as anchors. Generally the implementation is simpler.

Reviewer: Maybe note that this is the HDK's opinionated implementation of the anchor concept.

```

Retrieve the _names_ of all the second-level anchors for a top-level anchor by calling [`hdk::prelude::list_anchor_tags`](https://docs.rs/hdk/latest/hdk/hash_path/anchor/fn.list_anchor_tags.html):

```rust
use hdk::prelude::*;
use movie_integrity::*;

let all_first_letters = list_anchor_tags(LinkTypes::MovieByFirstLetterAnchor, "movies_by_first_letter");
```

Retrieve the _addresses_ of all the second-level anchors for a top-level anchor by calling [`hdk::prelude::list_anchor_addresses`](https://docs.rs/hdk/latest/hdk/hash_path/anchor/fn.list_anchor_addresses.html):

```rust
use hdk::prelude::*;
use movie_integrity::*;

let hashes_of_all_first_letters = list_anchor_addresses(LinkTypes::MovieByFirstLetterAnchor, "movies_by_first_letter");
```

Retrieve the _addresses_ of all the top-level anchors by calling [`hdk::prelude::list_anchor_addresses`](https://docs.rs/hdk/latest/hdk/hash_path/anchor/fn.list_anchor_addresses.html):

```rust
use hdk::prelude::*;
use movie_integrity::*;

let hashes_of_all_first_letters = list_anchor_addresses(LinkTypes::MovieByFirstLetterAnchor, "movies_by_first_letter");

> ```

Reviewer 2: Are these supposed to have the same example code?

Author: solved by deleting all mentions of `hash_path::Anchor`

> ## Reference

Reviewer: There are some opinionated path libraries that maybe we should link to here:
- https://github.com/holochain-open-dev/holochain-prefix-index
- https://github.com/holochain-open-dev/holochain-time-index
- https://github.com/holochain-open-dev/ranking-index
- https://github.com/holochain-open-dev/keyword-index

Author: I skipped the last two because I don't want to point to obsolete code. But yeah, great suggestion.

Reviewer: Unfortunately I think they're all outdated -- maybe we should leave them out for now and make an issue to include when they are updated.

## Retrieve links

Get all the _live_ (undeleted) links attached to a hash with the [`hdk::prelude::get_links`](https://docs.rs/hdk/latest/hdk/link/fn.get_links.html) function. The input is complicated, so use [`hdk::link::builder::GetLinksInputBuilder`](https://docs.rs/hdk/latest/hdk/link/builder/struct.GetLinksInputBuilder.html) to build it.

```rust
use hdk::prelude::*;
use movie_integrity::*;

let director_entry_hash = EntryHash::from_raw_36(vec![/* hash of Sergio Leone's entry */]);
let movies_by_director = get_links(
    GetLinksInputBuilder::try_new(director_entry_hash, LinkTypes::DirectorToMovie)?
        .get_options(GetStrategy::Network)
        .build()
)?;
let movie_entry_hashes = movies_by_director
    .iter()
    .filter_map(|link| link.target.into_entry_hash())
    .collect::<Vec<_>>();
```

If you want to filter the returned links by tag, pass some bytes to the input builder's `tag_prefix` method. You'll get back a vector of links whose tag starts with those bytes.

```rust
let movies_in_1960s_by_director = get_links(
    GetLinksInputBuilder::try_new(director_entry_hash, LinkTypes::DirectorToMovie)?
        .get_options(GetStrategy::Network)

>         ..tag_prefix("year:196".as_bytes().to_owned().into())

Reviewer: ```suggestion
        .tag_prefix("year:196".as_bytes().to_owned().into())
```

);
```

A link is considered ["dead"](/build/working-with-data/#deleted-dead-data) (deleted but retrievable if asked for explicitly) once its creation action has at least one delete-link action associated with it. As with entries, dead links can still be retrieved with [`hdk::prelude::get_details`](https://docs.rs/hdk/latest/hdk/prelude/fn.get_details.html) or [`hdk::prelude::get_link_details`](https://docs.rs/hdk/latest/hdk/link/fn.get_link_details.html) (see next section).

## Retrieve links

> Get all the _live_ (undeleted) links attached to a hash with the [`hdk::prelude::get_links`](https://docs.rs/hdk/latest/hdk/link/fn.get_links.html) function.

Reviewer: :+1:

);
```

!!! info Links are counted locally

> Currently `count_links` retrieves all links from the remote peer, then counts them locally. We're planning on changing this behavior so it counts links on the remote peer and sends the count, to save network traffic like you might expect.

Reviewer: wow I just searched MM and found the discussion about this. I did not realize this!

> Currently `count_links` retrieves all links from the remote peer, then counts them locally. We're planning on changing this behavior so it counts links on the remote peer and sends the count, to save network traffic like you might expect.

Reviewer: ```suggestion
Currently `count_links` retrieves all link hashes from the remote peer, then counts them locally. So it is less network traffic than a `get_links` request, but more network traffic than just sending an integer.
```

wow I just searched MM and found the discussion about this. I did not realize this!

Though actually I like it better than how I thought i was implemented since it does appear to merge your locally stored links with the received hashes and thus you can receive a links count integer that is less than your locally stored links. I don't know if we want to state we'll be refactoring it in this way.

Author: I learned something new too! didn't realise it was just the link _hashes_; that's a little better. I guess getting a remote count would potentially be less accurate, but if all I want is an unread badge or something like that, I'll take performance over precision (which is never guaranteed anyway).

While you can build this yourself, this is such a common pattern that the HDK implements it for you in the [`hdk::hash_path`](https://docs.rs/hdk/latest/hdk/hash_path/index.html) module. With it, you can create **paths**, which are hierarchies of anchors.

!!! info Avoiding DHT hot spots
Too many links on a single base address creates extra work for the peers responsible for it. Use a hierarchy of paths to split the links into appropriate 'buckets' and spread the work around. We'll give an example of that [below](#paths-in-depth).
!!!

### Scaffold a simple collection anchor

The scaffolding tool can create a 'collection', which is a one-level path that serves as an anchor for entries of a given type, along with all the functionality that creates and deletes links from that anchor to its entries:

```bash
hc scaffold collection
```

Follow the prompts to choose the entry type, name the link types and anchor, and define the scope of the collection, which can be either:

* all entries of type, or
* entries of type by author

### Paths in depth

> When you want to create more complex collections, you'll need to use the paths library directly.

Reviewer: ```suggestion
When you want to create more complex collections, you'll want to use the paths library directly.
```

let path_to_movies_starting_with_g = Path::from("movies_by_first_letter.g");
let links_to_movies_starting_with_g = get_links(
let links_to_movies_starting_with_g = get_links(
    // A path doesn't need to have a type in order to compute its hash.

>     GetLinksBuilder::try_new(path_to_movies_starting_with_g.path_entry_hash()?, LinkTypes::MovieByFirstLetter)?

Reviewer: ```suggestion
    GetLinksInputBuilder::try_new(path_to_movies_starting_with_g.path_entry_hash()?, LinkTypes::MovieByFirstLetter)?
```
